### PR TITLE
Anti-aliasing and better joining for the graph module

### PR DESCRIPTION
**Pull request by arch1t3cht30**
Two improvements to the graph shader module:

First off (commits 1 and 3), joining the two audio channels together in the middle instead of just clamping both of them down, which (to me) looks a lot better. In order to avoid sharp spikes in the middle we interpolate with quite a flat curve, which however results in a wide dent in the middle. A cleaner solution would be to join up both audio channels (after the fft) to just one sampler1d, then apply all the gravity and averaging algorithms and just render it as one whole graph instead of two separate sides, but I don't think that's possible in glsl only and did not want to change all of the glava source code.

Secondly (commit 2), adding an option for antialiasing the borders of the graph (by making each one-pixel-column (linearly, which looked best to me) fade to transparency until it reaches the highest adjacent column). This makes the graph module look way better to me, although it does require the xroot option. Since computations only happen next to the border of the graph, the performance impact should be marginal, but the optimizations did result in some slightly redundant/ugly shader code.
Since the anti-aliasing color is based on the color of the top of the respective column, this is compatible with the outline and highlight settings, but since the anti-aliasing only happens on the borders of the shape, and not on the seams between graph and outline, there might still be some hard edges with those settings on, so that the settings looks best with highlight and outline switched off.
In theory this could be improved even more by also considering the columns two steps to the left and right, but the result looked sufficient to me and my main goal was to get anti-aliasing working in the first place.

### CODE REVIEW

**Comment by arch1t3cht30** [comment]
Update: I added commit 4 which uses cubic interpolation when joining the channels instead of raising to a power of 1.8, which looks a lot smoother, does not create as big a dent, and does not cause spikes at the end of the dent. (In theory this could also be done when clamping down the ends of the graph but I only wanted to improve the center for now)

File: shaders/graph.glsl [hunk 1]
```diff
@@ -13,8 +13,17 @@
 #define DRAW_OUTLINE 0
 /* 1 to draw edge highlight, 0 to disable */
 #define DRAW_HIGHLIGHT 1
+/* Whether to anti-alias the border of the graph, creating a smoother curve.
+   This may have a small impact on performance. Note that this only anti-aliases
```
Comment (arch1t3cht30) [inline shaders/graph.glsl]: I put in the performance note just to be sure, but I did not really notice any performance drop on my machine, so you might consider leaving it out depending on how it runs for you.

**Comment by wacossusca34** [comment]
Anti-aliasing support is a welcome improvement. I see no issues with it by looking over the code, and the performance impact should be very minimal (very few samples), although it is quite simple in nature.

I'll also try the cubic interpolation you added for joining the channels and see how it renders.

> A cleaner solution would be to join up both audio channels (after the fft) to just one sampler1d, then apply all the gravity and averaging algorithms and just render it as one whole graph instead of two separate sides, but I don't think that's possible in glsl only and did not want to change all of the glava source code.

It's not possible in pure GLSL but this is a good thing to mention, as I can add support for requesting a single sampler2D for both combined channels. It would probably produce the best output without any work in the shader.

Thanks for the contribution and following the existing style and format for shader passes. I'll try to get around testing this properly when I'm not sick.

**Comment by wacossusca34** [comment]
@arch1t3cht30 I would prefer invert being implemented in the existing passes. An entire pass just to handle inverting the results seems a bit excessive; I would rather the case be handled in the aliasing shader.

> I added the cleanest fix to this that I could think of,

For GLSL, I do not expect 'clean' code. It's fairly commonplace to perform otherwise ugly optimizations (ie. manually unrolling loops) due to how most GPU architectures work, and my abundant use of preprocessor switches (for the sake of avoiding branching) doesn't help either. To write what I consider 'clean' code in GLSL would mean sacrificing performance, or switching to a language that transpiles to GLSL.

The only thing I expect to be clean is the user configuration files, hence why I have provided hex color macros (`#rrggbbaa` syntax) and support for re-defining macros.

**Comment by wacossusca34** [comment]
Another thing to note about performance: its very hard to gauge performance on a modern dedicated graphics card. They are much better at handling branches than GPUs from the past, and have absurd amounts of cores.

If you're in my boat and can't test on integrated graphics, another option is to use Xephyr to get a seperate Xorg server running on `llvmpipe`, and run GLava in that. This is a somewhat realistic way to simulate how GLava would run on poor integrated chipsets. It also rules out issues with the mesa graphics stack if your current hardware is proprietary `nvidia`.

**Comment by arch1t3cht30** [comment]
> I would prefer invert being implemented in the existing passes.

No problem, I did that now.
I will try the benchmarking tomorrow though.

One more thing I thought of - if adding another shader pass has a performance impact, is it possible to remove shader passes entirely through preprocessor directives, such that when `ANTI_ALIAS` is off (such that the pass would do nothing at all) the entire file is never even compiled and added into the pipeline by glava? I think the same would also work for the `OUTLINE` and `HIGHLIGHT` effects.

**Comment by arch1t3cht30** [comment]
That makes sense. It was just a thought to get out a little bit more performance.
I tested the anti-aliasing on Xephyr for two resolutions and got the following results:

- On 800x500: Without the third shader pass: 49 fps, with the third pass and anti-aliasing off: 45 fps, With anti-aliasing on: 33 fps
- On 1280x500: Without the third shader pass: 35 fps, with the third pass and anti-aliasing off: 32 fps, With anti-aliasing on: 22 fps

This gives a drop in framerate of about 9% for adding another shader pass, and another approximate 30% for enabling anti-aliasing, when inside Xephyr.
If it helps, this is on an Intel i3-2120 CPU (4 cores, 3.3 GHz)

**Comment by wacossusca34** [comment]
@arch1t3cht30 that's actually a fairly substantial amount of overhead (more than I expected). I think it's worth merging this, but then working towards adding support to disable shader passes depending on the configuration.

Since you are on Intel (as I am stuck with Ryzen), would you be able to test using integrated graphics?

Another thing worth looking at is adding some fixed, geometry-based modules instead of rendering everything in a fragment shader. Integrated and software rasterizing unfortunately is not well-suited to the fairly heavy shaders that GLava makes use of.

This of course would be limited to modules like `bars` and `graph`, mostly so that users without dedicated graphics hardware can afford to constantly run a visualizer in the background (as at the moment, this tends to only be the case for those who have a free GPU).

One of the benefits of this approach would be _much_ faster anti-aliasing by rendering the graph as a single polygon.

**Comment by wacossusca34** [comment]
After some of my own testing I couldn't find any other issues, so I have merged it. I still wish to add some way to disable shader passes.

> Another possible (but very dirty) solution would be to automatically disable a shader if it contains a compiler error

This is actually a very good idea if you use it with the standard GLSL `#error` macro, ie:

```GLSL
#if ANTI_ALIAS = 0
#error __disablestage
#endif
```

It would be fairly easy to strip the occurrences this way, and its far less ugly than inducing errors using undefined symbols.

I applaud your creativity, I may actually implement this.

**Comment by wacossusca34** [comment]
The (edited) code in my previous comment now works. While I agree it's a bit of a hack, I feel it's necessary for the sake of improving performance -- I also don't see any other elegant solutions that don't require me to write a preprocessor.

When a stage is disabled, the renderer simply 'skips' it and behaves as if it doesn't exist in the list of passes to perform -- such that if pass `2` is disabled, pass `3` will apply to the inputs from `1`.

At the moment it blindly scans error output for the string `__disablestage`. I would like to more intelligently scan for strings from the compiler output, but AFAIK there isn't really any standard format for how errors are formatted by the driver.

**Comment by wacossusca34** [comment]
@arch1t3cht30 I would like to eventually alter this implementation to instead use the alpha channel as that will blend while using `"native"` transparency, while still working with other modes that blend using OpenGL.

This is how the `circle` module handles aliasing, although it's worth mentioning that for `"native"`, you need an alpha premultiply step that is provided as a utility (due to how X11 handles window transparency):

```GLSL
/* final pass: */
#if ANTI_ALIAS == 0
#error __disablestage
#endif
#include ":util/premultiply.frag"
```

Note in the source of `premultiply.frag`, there is a special `#request nativeonly true` line that only enables the shader for `"native"` opacity. Since the `__disablestage` feature has been added, it might be a bit more uniform to now expose this as a macro instead.

If you would like to do so yourself, let me know, otherwise I'll leave it on my list of things to do.

**Comment by arch1t3cht30** [comment]
I'm not at home at the moment so it would take some time for me to get to it, so I'll have to leave it to you for now, sorry.

**Comment by wacossusca34** [comment]
Your code actually didn't need any changes to work with `"native"` opacity, just the aforementioned pass for premultiplying the `rgb` channels.

**Pull request merged**